lib/join/oraclejoin/oracle.go

```diff
@@ -0,0 +1,328 @@
+	}
+
+	// Unset RequestURI and set URL (necessary quirk of sending a request parsed by
+	// http.ReadRequest)
+	httpReq.RequestURI = ""
+	httpReq.URL.Scheme = "https"
+	httpReq.URL.Host = httpReq.Host
+	// If the HTTP client sees this header set to gzip it will expect the
+	// caller to handle gzip itself instead of decoding automatically.
+	delete(httpReq.Header, "Accept-Encoding")
+	return httpReq, nil
+}
+
+func validateRootCAReq(req *http.Request, instanceID string) error {
+	const rootCAPath = "/v1/instancePrincipalRootCACertificates"
+	if req.URL.Path != rootCAPath {
+		return trace.BadParameter("path must be %s, got %s", rootCAPath, req.URL.Path)
+	}
+	expectedRegion, err := regionFromInstanceID(instanceID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	expectedHostname := expectedRegion.Endpoint("auth")
+	if req.URL.Hostname() != expectedHostname {
+		return trace.BadParameter("hostname must be %s, got %s", expectedHostname, req.URL.Hostname())
+	}
+	return nil
+}
+
+// regionFromInstanceID returns an oci region for a given instance ID. It will
+// return an error if the region name is not recognized by the SDK. This is a
+// critical check since this region will be used to validate the endpoint that
+// will be used to fetch the regional root CAs. If a bogus region were allowed
+// then teleport could be tricked into making a request to
+// auth.<bogus-region-name>.oraclecloud.com with potential tricks in
+// <bogus-region-name>.
+func regionFromInstanceID(instanceID string) (common.Region, error) {
+	// Expected InstanceID format ocid1.instance.<realm>.[region][.future-use].<id>
+	chunks := strings.SplitN(instanceID, ".", 5)
+	if len(chunks) < 5 {
+		return "", trace.BadParameter("instance ID does not have expected format, got %s", instanceID)
+	}
+	regionShortName := chunks[3]
+	region := common.StringToRegion(regionShortName)
+	if _, err := region.RealmID(); err != nil {
+		// StringToRegion always returns something, RealmID will return an
+		// error if it's not a real region.
+		return "", trace.AccessDenied("unsupported region %s", regionShortName)
+	}
+	return region, nil
+}
+
+func executeRootCAReq(ctx context.Context, client utils.HTTPDoClient, req *http.Request) (*rootCAResp, error) {
+	req = req.WithContext(ctx)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, trace.Wrap(err, "sending request to Oracle API")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, trace.Errorf("HTTP request failed with status: %s", resp.Status)
+	}
+
+	body, err := io.ReadAll(io.LimitReader(resp.Body, teleport.MaxHTTPResponseSize))
+	if err != nil {
+		return nil, trace.Wrap(err, "reading response body")
+	}
+
+	var parsedResp rootCAResp
+	if err := json.Unmarshal(body, &parsedResp); err != nil {
+		return nil, trace.Wrap(err, "parsing response")
+	}
+	return &parsedResp, nil
+}
+
+type rootCAResp struct {
+	Certificates []string `json:"certificates"`
```

Comment:
Nit: Perhaps a brief comment here explaining what format we expect these certificates to be in? I'm assuming it's just base64 encoded DER ASN.1?

Reply:
added in [d0c9be8](https://github.com/gravitational/teleport/pull/60601/commits/d0c9be8615fb231db433663ba85f6ced85fe7005)